### DISCUSSION TITLE
How to generate statistics from my Commons Booking data?

### DISCUSSION THREAD

**Discussion by Splines** [wielebenwir/commonsbooking#1101] [Questions & Answers]
Need some statistical evaluation of your Commons Booking data, e.g. how many users booked one of your items per week or per weekday? Check out our own little Python solution (MIT-licensed) — brought to you by [inSPEYERed](https://inspeyered.de) — to generate a neat PDF file with all statistics combined:

| 📊  | [cb-statistics](https://github.com/inSPEYERed/cb-statistics) |
|---------------|:-------------------------|

**UPDATE: we now offer a [website](https://inspeyered.github.io/cb-statistics/) where you can simply drag and drop your exported CSV data from Commons Booking and get the plots right in your browser.**

**Comment by hansmorb** [comment]
Habe noch Probleme mit dem Skript, bekomme diesen Fehler:
```

 from . import config, server
  File "/home/hansm/.local/lib/python3.10/site-packages/frontend/server.py", line 24, in <module>
    app.mount(config.STATIC_ROUTE, StaticFiles(directory=config.STATIC_DIRECTORY), name=config.STATIC_NAME)
  File "/home/hansm/.local/lib/python3.10/site-packages/starlette/staticfiles.py", line 55, in __init__
    raise RuntimeError(f"Directory '{directory}' does not exist")
RuntimeError: Directory 'static/' does not exist
```

Wenn ich fitz deinstalliere läuft das Skript weiter, hängt aber bei 

`💠 Evaluate bookings per item
Traceback (most recent call last):
  File "/home/hansm/Documents/Freie Lasten/statistics/cb-statistics/./src/main.py", line 36, in <module>
    per_item_results = evaluate_bookings_per_item(bookings, items_map)
  File "/home/hansm/Documents/Freie Lasten/statistics/cb-statistics/src/evaluation.py", line 80, in evaluate_bookings_per_item
    item_name = items_map[item]
KeyError: 2788
`

**Reply by Splines** [reply]
Hi Hans, ja, pipreqs hat bei mir fälschlicherweise `fitz` eingebaut in die `requirements.txt`, das habe ich übersehen und nun fitz daraus entfernt, sodass da stattdessen nur noch `PyMuPDF` installiert wird.

Das Skript hat dann später bei `evaluate_bookings_per_item` gehangen, weil ich da manuell eine Map hardcodiert habe mit den Ids unserer Lastenräder, siehe [hier](https://github.com/inSPEYERed/cb-statistics/blob/1daf9db47016899368a6c05f07d80ed35ac1ef23/src/main.py#L20-L27). Habe das nun auch automatisiert, sodass das Skript hoffentlich bei dir laufen sollte. Wie im Readme beschrieben, musst du eventuell bei den Plots [hier](https://github.com/inSPEYERed/cb-statistics/blob/main/src/plot_diagrams.py#L7-L8) die Farben erweitern, falls du mehr als drei Jahre Daten betrachten solltest.

**Reply by hansmorb** [reply]
Alles klar, danke dir! Konnte das Skript das erste Mal erfolgreich durchlaufen lassen, sehr cool! Ich fände es auch noch sehr schön in der Zukunft Support für verschiedene Artikelkategorien zu haben. Gibt es einen Grund, warum ihr nicht die von CommonsBooking exportierbare CSV nutzen wolltet?

**Reply by Splines** [reply]
Waaas, es gibt eine CSV-Export-Funktion?! Die habe ich wohl übersehen, dann hätte ich mir den Aufwand mit dem umständlichen Parsen der Datenbanktabellen wohl sparen können 😅 Jetzt habe ich es auch [in eurer Doku gefunden](https://commonsbooking.org/docs/einstellungen-2/export/).
Ja, verschiedene Artikelkategorien, sind sicherlich noch interessant (den Use Case haben wir persönlich allerdings noch nicht). Meinst du dann so was wie Anzahl Buchungen pro Artikelkategorie?

**Reply by Splines** [reply]
@hansmorb Die Export-Funktion funktioniert [bei uns](https://inspeyerad.de) irgendwie nicht... Bekomme ein Gateway Timeout und die gesamte Seite ist für ca. 2-3 Minuten nicht mehr erreichbar.

**Reply by genevievecory** [reply]
Probier mal nur ein Monat zu exportieren, dann geht's wahrscheinlich... nach einem erfolgreichen kurzen Export kannst Du dann eine größere Menge anstossen... so hat's bei uns funktioniert.

**Reply by hansmorb** [reply]
So ist es! Die heruntergeladenen Zeitrahmen werden auch gecachet, wenn du mit einem kleineren Wert anfängst, dann sind die vorherigen Werte schon im Zwischenspeicher und du kannst danach größere Mengen exportieren. Der Prozess schlägt fehl, wenn deine PHP Einstellungen ein zu geringes memory_limit oder die max_execution_time überschritten ist. Das kannst du alles in der php.ini einstellen.

**Reply by Splines** [reply]
@genevievecory @hansmorb Super, vielen Dank euch, der Tipp mit "erst mal kleinere Ranges exportieren" hat geholfen. Zwischendurch ist es mal zu einem Bug gekommen, ich habe das Issue #1104 (mit Stacktrace) dazu aufgemacht. Ich probiere jetzt auch noch mal die PHP-Einstellungen zu ändern.

**Comment by hansmorb** [comment]
Und ein paar Sachen fallen mir noch auf: Es scheint so als werden ein paar Buchungen verschluckt. Kann ich dir mal nen Testdatensatz schicken?

**Reply by Splines** [reply]
Ja, gerne an info at splines.me
Ich habe die Buchungen gefiltert nach denen die `confirmed` sind, d.h. die cancelled fallen zum Beispiel raus, siehe [hier](https://github.com/inSPEYERed/cb-statistics/blob/main/src/evaluation.py#L22-L23).

**Comment by hansmorb** [comment]
Hey, coole Sache mit der Webseite. Wir haben auch intern gelegentlich https://pivottable.js.org/examples/local.html genutzt. Dort kann auch dynamisch aus den Elementen ein Pivot Tabelle erstellt werden.

**Reply by Splines** [reply]
Das ist ja auch cool, danke für den Tipp.

**Comment by hansmorb** [comment]
Irgendwie malen sich auf Firefox bei mir die Plots nicht, kannst du das mal ausprobieren?

**Reply by Splines** [reply]
Danke, sollte nun gefixed sein mit diesem Commit: https://github.com/inSPEYERed/cb-statistics/commit/407e444409d453c2164f6af8ea0f62f5119002d3

Es wäre klasse, wenn ihr für Datumsangaben ISO 8601 verwenden könntet, siehe [hier](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date/parse):

> As until ES5, parsing of strings was entirely implementation-dependent. After the ES5 spec, the [ISO 8601 format](https://tc39.es/ecma262/#sec-date-time-string-format) (YYYY-MM-DDTHH:mm:ss.sssZ) is explicitly specified, while other formats are still implementation-defined and may not work across all browsers.

Und nicht nur in ES5 ist das ja mittlerweile zum Standard geworden.

**Reply by hansmorb** [reply]
Gute Idee, ist im PR #1108
